tests/platform_tests/test_auto_negotiation.py

```diff
@@ -0,0 +1,356 @@
+import logging
+import pytest
+import random
+
+from natsort import natsorted
+from tests.common.config_reload import config_reload
+from tests.common.helpers.assertions import pytest_assert, pytest_require
+from tests.common.helpers.dut_ports import decode_dut_port_name
+from tests.common.utilities import wait_until
+from tests.platform_tests.link_flap.link_flap_utils import build_test_candidates
+
+logger = logging.getLogger(__name__)
+
+STATE_DB = 'STATE_DB'
+STATE_PORT_TABLE_TEMPLATE = 'PORT_TABLE|{}'
+STATE_PORT_FIELD_SUPPORTED_SPEEDS = 'supported_speeds'
+APPL_DB = 'APPL_DB'
+APPL_PORT_TABLE_TEMPLATE = 'PORT_TABLE:{}'
+ALL_PORT_WAIT_TIME = 60
+SINGLE_PORT_WAIT_TIME = 40
+PORT_STATUS_CHECK_INTERVAL = 10
+
+# To avoid getting candidate test ports again and again, use a global variable
+# to save all candidate test ports. 
+# Key: dut host object, value: a list of candidate ports tuple
+cadidate_test_ports = {}
+
+
+@pytest.fixture(scope='module', autouse=True)
+def recover_ports(duthosts, enum_dut_portname_module_fixture, fanouthosts):
+    """Module level fixture that automatically do following job:
+        1. Build global candidate test ports 
+        2. Save fanout port state before the test
+        3. Restor fanout and DUT after test
+
+    Args:
+        duthosts: DUT object
+        enum_dut_portname_module_fixture (str): DUT port name
+        fanouthosts: Fanout objects
+    """
+    global cadidate_test_ports
+    fanout_original_port_states = {}
+    dutname, portname = decode_dut_port_name(enum_dut_portname_module_fixture)
+    logger.info('Collecting existing port configuration for DUT and fanout...')
+    for duthost in duthosts:
+        if dutname == 'unknown' or dutname == duthost.hostname:
+            all_ports = build_test_candidates(duthost, fanouthosts, portname)
+            # Test all ports takes too much time (sometimes more than an hour), 
+            # so we choose 3 ports randomly as the cadidates ports
+            cadidate_test_ports[duthost] = random.sample(all_ports, 3)
```

Comment:
Will it happen in a high possibility that all the 3 ports are not testable? means fanout or cable to support AN?

Reply:
I never met it during my test.